python/amici/conserved_moieties.py

```diff
@@ -0,0 +1,862 @@
+                    pivots[order[j + 1]] = matrix[order[j + 1]][0]
+                else:
+                    pivots[order[j + 1]] = MAX
+
+    RSolutions = [[] for _ in range(num_species)]
+    RSolutions2 = [[] for _ in range(num_species)]
+    kernel_dim = 0
+
+    for i in range(num_species):
+        done = all(matrix[i][j] >= num_reactions
+                   for j in range(len(matrix[i])))
+        if done and len(matrix[i]):
+            for j in range(len(matrix[i])):
+                RSolutions[kernel_dim].append(matrix[i][j] - num_reactions)
+                RSolutions2[kernel_dim].append(matrix2[i][j])
+            kernel_dim += 1
+    del matrix, matrix2
+
+    matched = []
+    int_matched = []
+    cls_species_idxs = [[] for _ in range(num_species)]
+    cls_coefficients = [[] for _ in range(num_species)]
+
+    i2 = 0
+    for i in range(kernel_dim):
+        ok2 = True
+        for j in range(len(RSolutions[i])):
+            if RSolutions2[i][j] * RSolutions2[i][0] < 0:
+                ok2 = False
+            if not matched or all(
+                    cur_matched != RSolutions[i][j] for cur_matched in
+                    matched
+            ):
+                matched.append(RSolutions[i][j])
+        if ok2 and len(RSolutions[i]):
+            min_value = MAX
+            for j in range(len(RSolutions[i])):
+                cls_species_idxs[i2].append(RSolutions[i][j])
+                cls_coefficients[i2].append(abs(RSolutions2[i][j]))
+                min_value = min(min_value, abs(RSolutions2[i][j]))
+                if not int_matched or all(
+                        cur_int_matched != cls_species_idxs[i2][j]
+                        for cur_int_matched in int_matched
+                ):
+                    int_matched.append(cls_species_idxs[i2][j])
+            for j in range(len(cls_species_idxs[i2])):
+                cls_coefficients[i2][j] /= min_value
+            i2 += 1
+    int_kernel_dim = i2
+
+    assert int_kernel_dim <= kernel_dim
+    assert len(cls_species_idxs) == len(cls_coefficients), \
+        "Inconsistent number of conserved quantities in coefficients and " \
+        "species"
+    return (kernel_dim, matched, int_kernel_dim, int_matched, cls_species_idxs,
+            cls_coefficients)
+
+
+def _fill(
+        stoichiometric_list: Sequence[float],
+        matched: Sequence[int],
+        num_species: int
+) -> Tuple[List[List[int]], List[List[int]], List[int]]:
+    """Construct interaction matrix
+
+    Construct the interaction matrix out of the given stoichiometric matrix
+    :math:`S`.
+
+    :param stoichiometric_list:
+        the stoichiometric matrix given as a flat list
+    :param matched:
+        found and independent moiety conservation laws (MCL)
+    :param num_species:
+        number of rows in :math:`S`
+    :returns:
+        interactions of metabolites and reactions, and matrix of interaction
+    """
+    dim = len(matched)
+    MIN = 1e-9
```

Comment:
my feeling is that MIN/MAX should be consistent across the different functions?

Reply:
I tend to agree.